chain/events/state/predicates.go

```diff
@@ -135,3 +140,123 @@ func (sp *StatePredicates) DealStateChangedForIDs(dealIds []abi.DealID) DiffDeal
 		return false, nil, nil
 	}
 }
+
+type DiffMinerActorStateFunc func(ctx context.Context, oldState *miner.State, newState *miner.State) (changed bool, user UserData, err error)
+
+func (sp *StatePredicates) OnMinerActorChange(minerAddr address.Address, diffMinerActorState DiffMinerActorStateFunc) DiffFunc {
+	return sp.OnActorStateChanged(minerAddr, func(ctx context.Context, oldActorStateHead, newActorStateHead cid.Cid) (changed bool, user UserData, err error) {
+		var oldState miner.State
+		if err := sp.cst.Get(ctx, oldActorStateHead, &oldState); err != nil {
+			return false, nil, err
+		}
+		var newState miner.State
+		if err := sp.cst.Get(ctx, newActorStateHead, &newState); err != nil {
+			return false, nil, err
+		}
+		return diffMinerActorState(ctx, &oldState, &newState)
+	})
+}
+
+type MinerSectorChanges struct {
+	Added    []miner.SectorOnChainInfo
+	Extended []SectorExtensions
+	Removed  []miner.SectorOnChainInfo
+}
+
+type SectorExtensions struct {
+	From miner.SectorOnChainInfo
+	To   miner.SectorOnChainInfo
+}
+
+func (sp *StatePredicates) OnMinerSectorChange() DiffMinerActorStateFunc {
+	return func(ctx context.Context, oldState, newState *miner.State) (changed bool, user UserData, err error) {
+		ctxStore := &contextStore{
+			ctx: context.TODO(),
+			cst: sp.cst,
+		}
+
+		sectorChanges := &MinerSectorChanges{
+			Added:    []miner.SectorOnChainInfo{},
+			Extended: []SectorExtensions{},
+			Removed:  []miner.SectorOnChainInfo{},
+		}
+
+		// no sector changes
+		if oldState.Sectors == newState.Sectors {
+			return false, nil, nil
+		}
+
+		oldSectors, err := adt.AsArray(ctxStore, oldState.Sectors)
+		if err != nil {
+			return false, nil, err
+		}
+
+		newSectors, err := adt.AsArray(ctxStore, newState.Sectors)
+		if err != nil {
+			return false, nil, err
+		}
+
+		var osi miner.SectorOnChainInfo
+
+		// find all sectors that were extended or removed
+		if err := oldSectors.ForEach(&osi, func(i int64) error {
+			var nsi miner.SectorOnChainInfo
+			found, err := newSectors.Get(uint64(osi.Info.SectorNumber), &nsi)
+			if err != nil {
+				return err
+			}
+			if !found {
+				sectorChanges.Removed = append(sectorChanges.Removed, osi)
+				return nil
+			}
+
+			if nsi.Info.Expiration != osi.Info.Expiration {
```

Comment:
This can also happen when a CC sector is upgraded - then the Expiration field is set to the nearest proving period boundary

Reply:
Can the DealID's in a sector ever change? (And is this just a passing comment or a request for change?)

Reply:
> Can the DealID's in a sector ever change?

No, sectorid<->commr/deals is a 1:1 mapping (except in some weird/crafted reorg cases). When a sector is upgraded it gets a new SectorID